### ISSUE TITLE
6-8 minutes waiting on /api/contents

### ISSUE THREAD

**Issue by dodysw** [jupyter/notebook#1888]
Linux RHEL 6.7 within companies set of VM, latest Jupyter as from "conda update --all", Python 3.5, and Python 3.5 kernel.

Yesterday when suddenly whenever I open or create a new notebook, Jupyter stuck/hung. In the console log, all I can see is these...:
```
[D 10:58:55.114 NotebookApp] Found kernel python3 in /opt/logs/sandbox/conda/share/jupyter/kernels
[D 10:58:55.128 NotebookApp] 304 GET /api/kernelspecs (x.x.x.x) 38.19ms

**<after 7 minutes>**

[D 11:06:24.130 NotebookApp] 200 GET /api/contents/2016-10-27-consumer.ipynb?type=notebook&_=1478883534723 (x.x.x.x) 449000.68ms
...
[D 11:06:25.168 NotebookApp] Starting kernel: ['/opt/logs/sandbox/conda/bin/python', '-m', 'ipykernel', '-f', '/home/xxx/.local/share/jupyter/runtime/kernel-d6bd7421-7664-46a1-86eb-39d2ac515966.json']
[D 11:06:25.175 NotebookApp] Connecting to: tcp://127.0.0.1:55392
[I 11:06:25.176 NotebookApp] Kernel started: d6bd7421-7664-46a1-86eb-39d2ac515966

# Then finally shows up until 7.5 minutes.
# When executing a cell:
# nothing for 6 minutes, then....
[D 11:14:39.551 NotebookApp] 200 PUT /api/contents/2016-10-27-consumer.ipynb (x.x.x.x) 375126.00ms
```

Pretty sure this is related to the environment, possibly some antivirus, iptables, or something behind the scene was implemented automatically as standard policy, I don't have root access so can't find out exactly what was going on. 

Is there a way for me to know more about what /api/contents is actually waiting for to help me know a little more about what was actually going on with tornado/ipython/jupyter?

**Comment by takluyver** [comment]
Is there anything unusual about that notebook (e.g. is it particularly big)? Can you find anything that changed yesterday that could have affected this (e.g. updates installed)?

The code that handler is using to retrieve the notebook is [here](https://github.com/jupyter/notebook/blob/2bbaaa5cde5c2dc1a42042db02c6cc4025a24cee/notebook/services/contents/filemanager.py#L331) if you want to break down exactly which operation is slow.

**Comment by dodysw** [comment]
There is nothing unusual. This happened too even when creating new (empty) notebook. Browsing the file tree in Jupyter is fine. Opening terminal is fine. The reason I'm thinking it's the environment is because it didn't happen in my other Linux VMs not managed by corporate. 

Thanks for pointing the entry point, will get back to this thread my discoveries after putting print statements...

**Comment by dodysw** [comment]
It was stucked at [this line](https://github.com/jupyter/nbformat/blob/1eaa2e89381ec9a345bc39e4727ea1b86218733d/nbformat/sign.py#L252
)

really surprising, why would checking if an object is null would take a long time. Continuing with digging....

**Comment by takluyver** [comment]
Getting `self.db` for the first time will run [this function](https://github.com/jupyter/nbformat/blob/1eaa2e89381ec9a345bc39e4727ea1b86218733d/nbformat/sign.py#L132) which opens an sqlite database. We've had some issues with sqlite when on an NFS filesystem - is your home dir on NFS? Is `~/.local/share/jupyter/nbsignatures.db` getting big?

You can bypass the file access for this database by configuring `NotebookNotary.db_file=':memory:'`. This will mean all notebooks are untrusted, but if they load quickly, that's an improvement.

**Comment by dodysw** [comment]
Yes you are correct! The home dir is an NFS mount. I was thinking Jupyter only uses ~/.jupyter to store user files. The moment I saw the .db was an sqlite file and your reply came that it was stored in homedir, it is a problem. Pretty sure sqlite doesn't work in NFS.

Thanks for helping me with this and the work around. We can close this issue.

Tldr. If home directory is an NFS mount, Jupyter may not work as expected.

On November 11, 2016 12:16:44 PM CST, Thomas Kluyver <notifications@github.com> wrote:
>Getting `self.db` for the first time will run [this
>function](https://github.com/jupyter/nbformat/blob/1eaa2e89381ec9a345bc39e4727ea1b86218733d/nbformat/sign.py#L132)
>which opens an sqlite database. We've had some issues with sqlite when
>on an NFS filesystem - is your home dir on NFS? Is
>`~/.local/share/jupyter/nbsignatures.db` getting big?
>
>You can bypass the file access for this database by configuring
>`NotebookNotary.db_file=':memory:'`. This will mean all notebooks are
>untrusted, but if they load quickly, that's an improvement.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/jupyter/notebook/issues/1888#issuecomment-260020301

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by takluyver** [comment]
We are planning to move away from sqlite for the notebook signatures database - I'm hoping to have some time to work on that in the couple of days. #1782 will track that.

We also use an sqlite database for IPython history; you may have some trouble with that too. To avoid that, look up `HistoryAccessor.hist_file` in [this page](http://ipython.readthedocs.io/en/stable/config/options/terminal.html).

**Comment by dodysw** [comment]
You were right again on the IPython history, it was having the same problem too when starting IPython. Also, the size of sqlite db in my case is 380KB for IPython history.sqlite, and 63KB for nbsignatures.db.

Just to summarize the work arounds related to issues with Jupyter/IPython hanging/not-responding/stuck for 5 minutes when starting/opening/creating notebook due to sqlite DB file on NFS mount:
- IPython: pass `--HistoryAccessor.hist_file` parameter when running IPython to a different path on a non-NFS mount e.g. `ipython --HistoryManager.hist_file=/tmp/ipython_hist.sqlite`, or set environment variable `IPYTHONDIR` to a non-NFS mount directory. E.g. `export IPYTHONDIR=/opt/appserver/sandbox/.ipython`
- Jupyter: Adding NotebookNotary.db_file=':memory:' in `$HOME/.jupyter/jupyter_notebook_config.py` which creates the sqlite database in memory, so will not persist when Jupyter process restarted. Or moving data directory to non-NFS partition by setting `JUPYTER_DATA_DIR` environment variable before running Jupyter. E.g. 
`export JUPYTER_DATA_DIR=/opt/appserver/sandbox/.local/share`

The following work around from [the documentation](https://github.com/jupyter/notebook/blob/715a6c02db0f6ad639d63e4f6fdff0cc63b91ec2/docs/source/security.rst) doesn't seem to work, when starting Jupyter, log said data_dir was not recognized as a valid configuration:

> To share a signatures database among users, you can configure:
> 
> c.NotebookNotary.data_dir = "/path/to/signature_dir"
> to specify a non-default path to the SQLite database (of notebook hashes, essentially). 

I think that instruction should be replaced with setting JUPYTER_DATA_DIR environment variable instead (from looking at [this line code](https://github.com/jupyter/jupyter_core/blob/master/jupyter_core/paths.py#L51)).
